[Bug]: Cannot find module '@node-rs/argon2-linux-arm64-gnu'

### Is there an existing issue for this?

- [X] I have searched the existing issues

### OS/Web Information

-  Web Browser: Firefox 96.0.3 (64-bit)
-  Local OS: macOS 10.14.6 (Mojave)
-  Remote OS: [Debian bullseye](https://hub.docker.com/_/debian)
-  Remote Architecture: aarch64
-  `code-server --version`: code-server 4.0.2 5cdfe74686aa73e023f8354a9a6014eb30caa7dd


### Steps to Reproduce

1.  `git clone https://gitlab.b-data.ch/jupyterlab/julia/docker-stack.git`
2.  `cd docker-stack/ver`
3.  `docker build -t julia-ver -f latest.Dockerfile .`

### Expected

Successful build

### Actual

https://gitlab.b-data.ch/jupyterlab/r/r-ver/-/jobs/20883

### Logs

```
[...]

Error: Cannot find module '@node-rs/argon2-linux-arm64-gnu'
Require stack:
- /opt/code-server/node_modules/@node-rs/argon2/index.js
- /opt/code-server/out/node/util.js
- /opt/code-server/out/node/cli.js
- /opt/code-server/out/node/entry.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:889:15)
    at Function.Module._load (internal/modules/cjs/loader.js:745:27)
    at Module.require (internal/modules/cjs/loader.js:961:19)
    at require (internal/modules/cjs/helpers.js:92:18)
    at Object.<anonymous> (/opt/code-server/node_modules/@node-rs/argon2/index.js:172:31)
    at Module._compile (internal/modules/cjs/loader.js:1072:14)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1101:10)
    at Module.load (internal/modules/cjs/loader.js:937:32)
    at Function.Module._load (internal/modules/cjs/loader.js:778:12)
    at Module.require (internal/modules/cjs/loader.js:961:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/opt/code-server/node_modules/@node-rs/argon2/index.js',
    '/opt/code-server/out/node/util.js',
    '/opt/code-server/out/node/cli.js',
    '/opt/code-server/out/node/entry.js'
  ]
}

[...]
```

### Screenshot/Video

_No response_

### Does this issue happen in VS Code?

- [X] I cannot reproduce this in VS Code.

### Are you accessing code-server over HTTPS?

- [X] I am using HTTPS.

### Notes

Related issues:

*  https://github.com/coder/code-server/issues/4422
*  https://github.com/coder/code-server/pull/4733

Also the armhf build..

```
app/node_modules/@node-rs/argon2/index.js:201
    throw loadError
    ^

Error: Cannot find module '@node-rs/argon2-linux-arm-gnueabihf'
Require stack:
```

--- a/ci/build/build-standalone-release.sh
+++ b/ci/build/build-standalone-release.sh
@@ -1,6 +1,10 @@
 #!/usr/bin/env bash
 set -euo pipefail
 
+# This is due to an upstream issue with RHEL7/CentOS 7 comptability with node-argon2
+# See: https://github.com/cdr/code-server/pull/3422#pullrequestreview-677765057
+export npm_config_build_from_source=true
+
 main() {
   cd "$(dirname "${0}")/../.."
 

--- a/ci/build/npm-postinstall.sh
+++ b/ci/build/npm-postinstall.sh
@@ -18,6 +18,9 @@ detect_arch() {
 }
 
 ARCH="${NPM_CONFIG_ARCH:-$(detect_arch)}"
+# This is due to an upstream issue with RHEL7/CentOS 7 comptability with node-argon2
+# See: https://github.com/cdr/code-server/pull/3422#pullrequestreview-677765057
+export npm_config_build_from_source=true
 
 main() {
   # Grabs the major version of node from $npm_config_user_agent which looks like

--- a/package.json
+++ b/package.json
@@ -85,7 +85,7 @@
   },
   "dependencies": {
     "@coder/logger": "1.1.16",
-    "@node-rs/argon2": "^1.0.5",
+    "argon2": "^0.28.0",
     "compression": "^1.7.4",
     "cookie-parser": "^1.4.5",
     "env-paths": "^2.2.0",

--- a/src/node/util.ts
+++ b/src/node/util.ts
@@ -166,11 +166,7 @@ export const isHashMatch = async (password: string, hash: string) => {
   if (password === "" || hash === "" || !hash.startsWith("$")) {
     return false
   }
-  try {
-    return await argon2.verify(hash, password)
-  } catch (error: any) {
-    throw new Error(error)
-  }
+  return await argon2.verify(hash, password)
 }
 
 /**

--- a/src/node/util.ts
+++ b/src/node/util.ts
@@ -1,5 +1,4 @@
-import { logger } from "@coder/logger"
-import * as argon2 from "@node-rs/argon2"
+import * as argon2 from "argon2"
 import * as cp from "child_process"
 import * as crypto from "crypto"
 import envPaths from "env-paths"
@@ -170,8 +169,7 @@ export const isHashMatch = async (password: string, hash: string) => {
   try {
     return await argon2.verify(hash, password)
   } catch (error: any) {
-    logger.error(error)
-    return false
+    throw new Error(error)
   }
 }
 

--- a/test/package.json
+++ b/test/package.json
@@ -8,6 +8,7 @@
     "@types/node-fetch": "^2.5.8",
     "@types/supertest": "^2.0.11",
     "@types/wtfnode": "^0.7.0",
+    "argon2": "^0.28.0",
     "jest": "^27.3.1",
     "jest-fetch-mock": "^3.0.3",
     "jsdom": "^16.4.0",
@@ -19,6 +20,7 @@
   },
   "resolutions": {
     "ansi-regex": "^5.0.1",
+    "argon2/@mapbox/node-pre-gyp/tar": "^6.1.9",
     "set-value": "^4.0.1",
     "tmpl": "^1.0.5",
     "path-parse": "^1.0.7",

--- a/test/unit/node/util.test.ts
+++ b/test/unit/node/util.test.ts
@@ -141,16 +141,16 @@ describe("isHashMatch", () => {
     const actual = await util.isHashMatch(password, _hash)
     expect(actual).toBe(false)
   })
-  it("should return false if the hash doesn't start with a $", async () => {
+  it("should return false and not throw an error if the hash doesn't start with a $", async () => {
     const password = "hellowpasssword"
     const _hash = "n2i$v=19$m=4096,t=3,p=1$EAoczTxVki21JDfIZpTUxg$rkXgyrW4RDGoDYrxBFD4H2DlSMEhP4h+Api1hXnGnFY"
+    expect(async () => await util.isHashMatch(password, _hash)).not.toThrow()
     expect(await util.isHashMatch(password, _hash)).toBe(false)
   })
-  it("should return false if the password and hash don't match", async () => {
+  it("should reject the promise and throw if error", async () => {
     const password = "hellowpasssword"
     const _hash = "$ar2i"
-    const actual = await util.isHashMatch(password, _hash)
-    expect(actual).toBe(false)
+    expect(async () => await util.isHashMatch(password, _hash)).rejects.toThrow()
   })
 })
 

